File: ios/Cove/OnboardingManager.swift
import Foundation
import Observation

extension WeakReconciler: OnboardingManagerReconciler where Reconciler == OnboardingManager {}

@Observable
final class OnboardingManager: AnyReconciler, OnboardingManagerReconciler, @unchecked Sendable {
    @ObservationIgnored let rust: RustOnboardingManager
    @ObservationIgnored private let rustBridge = DispatchQueue(
        label: "cove.onboarding.rustbridge", qos: .userInitiated
    )
    let app: AppManager
    var state: OnboardingState
    var isComplete = false

    typealias Message = OnboardingReconcileMessage

    init(app: AppManager) {
        self.app = app
        let rust = RustOnboardingManager()
        self.rust = rust
        self.state = rust.state()
        rust.listenForUpdates(reconciler: WeakReconciler(self))
    }

    func dispatch(_ action: OnboardingAction) {
        rustBridge.async { [rust] in
            rust.dispatch(action: action)
        }
    }

    func reconcile(message: OnboardingReconcileMessage) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            switch message {
            case let .step(step):
                state.step = step
            case let .branch(branch):
                state.branch = branch
            case let .createdWords(words):
                state.createdWords = words
            case let .cloudBackupEnabled(enabled):
                state.cloudBackupEnabled = enabled
            case let .secretWordsSaved(saved):
                state.secretWordsSaved = saved
            case let .cloudRestoreState(cloudRestoreState):
                state.cloudRestoreState = cloudRestoreState
            case let .cloudRestoreIssueChanged(cloudRestoreIssue):
                state.cloudRestoreIssue = cloudRestoreIssue
            case let .cloudRestoreProviderHintChanged(cloudRestoreProviderHint):
                state.cloudRestoreProviderHint = cloudRestoreProviderHint
            case let .shouldOfferCloudRestore(shouldOfferCloudRestore):
                state.shouldOfferCloudRestore = shouldOfferCloudRestore
            case let .cloudRestoreAlertVisible(cloudRestoreAlertVisible):
                state.cloudRestoreAlertVisible = cloudRestoreAlertVisible
            case let .restoreStateChanged(restoreState):
                state.restoreState = restoreState
            case let .errorChanged(error):
                state.error = error
            case .complete:
                isComplete = true
            }
        }
    }

    func reconcileMany(messages: [OnboardingReconcileMessage]) {
        messages.forEach { reconcile(message: $0) }
    }
}

File: android/app/src/main/java/org/bitcoinppl/cove/OnboardingManager.kt
package org.bitcoinppl.cove

import android.util.Log
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import java.io.Closeable
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.bitcoinppl.cove_core.OnboardingAction
import org.bitcoinppl.cove_core.OnboardingManagerReconciler
import org.bitcoinppl.cove_core.OnboardingReconcileMessage
import org.bitcoinppl.cove_core.OnboardingState
import org.bitcoinppl.cove_core.RustOnboardingManager
import org.bitcoinppl.cove_core.types.WalletId

internal data class OnboardingSnapshot(
    val state: OnboardingState,
    val isComplete: Boolean,
)

internal fun reduceOnboardingSnapshot(
    snapshot: OnboardingSnapshot,
    message: OnboardingReconcileMessage,
): OnboardingSnapshot =
    when (message) {
        is OnboardingReconcileMessage.Step -> snapshot.copy(state = snapshot.state.copy(step = message.v1))
        is OnboardingReconcileMessage.Branch -> snapshot.copy(state = snapshot.state.copy(branch = message.v1))
        is OnboardingReconcileMessage.CreatedWords -> snapshot.copy(state = snapshot.state.copy(createdWords = message.v1))
        is OnboardingReconcileMessage.CloudBackupEnabled -> snapshot.copy(state = snapshot.state.copy(cloudBackupEnabled = message.v1))
        is OnboardingReconcileMessage.SecretWordsSaved -> snapshot.copy(state = snapshot.state.copy(secretWordsSaved = message.v1))
        is OnboardingReconcileMessage.CloudRestoreState -> snapshot.copy(state = snapshot.state.copy(cloudRestoreState = message.v1))
        is OnboardingReconcileMessage.CloudRestoreMessageChanged -> snapshot.copy(state = snapshot.state.copy(cloudRestoreMessage = message.v1))
        is OnboardingReconcileMessage.ShouldOfferCloudRestore -> snapshot.copy(state = snapshot.state.copy(shouldOfferCloudRestore = message.v1))
        is OnboardingReconcileMessage.ErrorMessageChanged -> snapshot.copy(state = snapshot.state.copy(errorMessage = message.v1))
        is OnboardingReconcileMessage.Complete -> snapshot.copy(isComplete = true)
    }

internal interface OnboardingRustHandle : Closeable {
    fun state(): OnboardingState

    fun dispatch(action: OnboardingAction)

    fun listenForUpdates(reconciler: OnboardingManagerReconciler)

    fun currentWalletId(): WalletId?
}

private class RealOnboardingRustHandle(
    private val rust: RustOnboardingManager = RustOnboardingManager(),
) : OnboardingRustHandle {
    override fun state(): OnboardingState = rust.state()

    override fun dispatch(action: OnboardingAction) {
        rust.dispatch(action)
    }

    override fun listenForUpdates(reconciler: OnboardingManagerReconciler) {
        rust.listenForUpdates(reconciler)
    }

    override fun currentWalletId(): WalletId? = rust.currentWalletId()

    override fun close() {
        rust.close()
    }
}

@Stable
class OnboardingManager internal constructor(
    val app: AppManager,
    private val rust: OnboardingRustHandle = RealOnboardingRustHandle(),
    mainDispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
    rustDispatcher: CoroutineDispatcher = Dispatchers.IO,
) : OnboardingManagerReconciler, Closeable {
    private val mainScope = CoroutineScope(SupervisorJob() + mainDispatcher)
    private val rustScope = CoroutineScope(SupervisorJob() + rustDispatcher)
    private val isClosed = AtomicBoolean(false)

    var state by mutableStateOf(rust.state())
        private set

    var isComplete by mutableStateOf(false)
        private set

    init {
        rust.listenForUpdates(this)
    }

    fun dispatch(action: OnboardingAction) {
        rustScope.launch {
            runCatching { rust.dispatch(action) }
                .onFailure { error ->
                    val actionType = action::class.simpleName ?: "Unknown"
                    Log.e(TAG, "onboarding action failed: $actionType", error)
                }
        }
    }

    fun currentWalletId(): WalletId? = rust.currentWalletId()

    override fun reconcile(message: OnboardingReconcileMessage) {
        mainScope.launch {
            val nextSnapshot = reduceOnboardingSnapshot(OnboardingSnapshot(state = state, isComplete = isComplete), message)
            state = nextSnapshot.state
            isComplete = nextSnapshot.isComplete
        }
    }

    override fun close() {
        if (!isClosed.compareAndSet(false, true)) return
        mainScope.cancel()
        rustScope.cancel()
        rust.close()
    }

    companion object {
        private const val TAG = "OnboardingManager"
    }
}
